### PR TITLE
Refactor part2

### LINKED ISSUE #544: Part 2: (NEW PREDICTION TASK API) Refactoring all the prediction tasks to use the convention required by the Trainer class:i.e. compute loss inside the forward method + return dict with `{"loss", "labels" and "predictions"}` during training/evaluation mode.
This part of refactoring includes 4 parts:

- Update base PredictionTask class:
      ⁃	Add the `targets` argument to the forward method  
      ⁃	Move compute_loss() inside the forward method 
      ⁃	Return output (dict with the three keys or  torch.Tensor) based on training and testing flags
      ⁃	Update calculate metrics to pass the `targets` to the forward call 

- Update Head and Model classes to support the new convention in their forward method call  + calculate_metrics

- Update the fit method  **[Done]**:  loss is computed inside the forward call + add flag `compute_metrics=True` to control whether to compute metrics during training or not.  Replace the `compute_loss`  call `loss = self.compute_loss(x, y)` by : 
```
outputs = self(x, y, training=True)
loss = outputs['loss']
if compute_metrics=True: 
    self.calculate_metrics(outputs['predictions'], outputs['labels'])
```

- Update the failing unit tests

### PR DESCRIPTION
**Pull request by nzarif**
Fixes #544 

### Goals :soccer:
This part of refactoring includes 4 parts:

- Update base PredictionTask class:
- Update Head and Model classes to support the new convention in their forward method call  + calculate_metrics
- Update the fit method  **[in progress]**:  loss is computed inside the forward call + add flag `compute_metrics=True` to control whether to compute metrics during training or not.  Replace the `compute_loss`  call `loss = self.compute_loss(x, y)` by : 
```
outputs = self(x, y, training=True)
loss = outputs['loss']
if compute_metrics=True: 
    self.calculate_metrics(outputs['predictions'], outputs['labels'], mode='train', forward=False, call_body=False)
```

- Update the failing unit tests

### Testing Details :mag:
Run the torch unit tests and you will see they pass. You can also try with [integration tests](https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/main/ci/test_integration.sh).

### CODE REVIEW

**Review by sararb — COMMENTED** [review]
Thank you for your PR!
- I pulled it locally together with the changes of PR #551 to run the CI integration test: all the tests passed and the evaluation scores didn't drop. so we can trust that the changes made are not affecting the model's performance. 
- I left some minor comments to leverage the new output API in the unit-tests
- I noticed that the [examples/tutorial/03-Session-based-recsys.ipynb](https://github.com/NVIDIA-Merlin/Transformers4Rec/pull/546/files#diff-ede55fb81e11c9acb228c321bd93dd9e6b8c937846d2169830f61d65b953e88a) has been modified but I couldn't find what are the changes made. So could you specify the changes in the PR's description?

File: tests/torch/model/test_head.py [hunk 1]
```diff
@@ -108,9 +109,12 @@ def test_item_prediction_head(torch_yoochoose_tabular_transformer_features, torc
     body = tr.SequentialBlock(input_module, tr.MLPBlock([64]))
     head = tr.Head(body, tr.NextItemPredictionTask(), inputs=input_module)
 
-    outputs = head(body(torch_yoochoose_like))
+    outputs = head(body(torch_yoochoose_like, training=True))
 
-    assert outputs.size()[-1] == input_module.categorical_module.item_embedding_table.num_embeddings
+    assert (
+        list(outputs.values())[0].size()[-1]
```
[the remark this answers is not in the document]
Comment (nzarif) [reply tests/torch/model/test_head.py:115]: It is not a dict of {'prediction':...'labels':...'loss':...} but it is a dict of {'task_name':Tenosr(prediction)}. The reason to this is that if Head gets rid of task_name and the Model has multiple heads to aggregate, Model won't be able to do aggregation. So Head must preserve task_name even if there is only 1 task. Considering user is not supposed to use the output of Head directly for serving, I think it is okay for Head not to return a Tensor for prediction. Model will take care of turning this dict to Tensor if len()==1.

File: transformers4rec/torch/model/base.py [hunk 2]
```diff
@@ -174,53 +186,29 @@ def child_name(self, name):
...
+            else:
+                raise ValueError("target_name can not be None, specify a name.")
 
         outputs = {}
         if forward:
-            predictions = self(predictions, training=training, testing=testing)
+            predictions = self(predictions, targets=targets, training=training, testing=testing)[
+                "predictions"
+            ]
+        if isinstance(predictions, dict):
```
[the remark this answers is not in the document]
Comment (nzarif) [reply transformers4rec/torch/model/base.py:210]: Yes, when we `Head.calculate_metrics` is called on this [line](https://github.com/NVIDIA-Merlin/Transformers4Rec/blob/ec92eb6500dad7a8e6222523ff54f89c80e196e6/transformers4rec/torch/model/base.py#L438) it receives `body_outputs` as predictions and that is shaped as: 
`x = {'classification/binary_classification_task': tensor([0.4769, 0.4737, 0.4775, 0.4768, 0.4794, 0.4741, 0.4749, 0.4809, ...-2.5440e-01, -2.7638e-01, -2.0624e-01, -3.4702e-01,  8.2734e-02, -2.0389e-01],
       grad_fn=<ViewBackward0>)}`

**Comment by nzarif** [comment]
> Thank you for your PR!
> 
> * I pulled it locally together with the changes of PR #551 to run the CI integration test: all the tests passed and the evaluation scores didn't drop. so we can trust that the changes made are not affecting the model's performance.
> * I left some minor comments to leverage the new output API in the unit-tests
> * I noticed that the [examples/tutorial/03-Session-based-recsys.ipynb](https://github.com/NVIDIA-Merlin/Transformers4Rec/pull/546/files#diff-ede55fb81e11c9acb228c321bd93dd9e6b8c937846d2169830f61d65b953e88a) has been modified but I couldn't find what are the changes made. So could you specify the changes in the PR's description?

The thing changed in Notebooks is 'hf_format' being removed as far as I remember. There was a merge conflict that is why that notebook was changed both in this pr and part1 pr. Removing hf_format but not removing the space after it had caused a merge conflict!

**Review by sararb — APPROVED** [review]

**Review by gabrielspmoreira — APPROVED** [review]
